## 章節 3：估計 (Estimation)

### 3.1 骰子問題

現在有一個盒子裝了4面、6面、8面、12面以及20面的骰子。如果你曾經玩過龍與地下城(Dungeons & Dragons)你就知道我在說什麼。

假設我從盒子裡隨便選了一個骰子，擲它，結果為6。請問每個骰子被擲的機率各是多少？

我邊我建議一個三步驟策略來解決這樣的問題。

1. 找一個方式表示假設組
1. 找一個方式表示資料
1. 撰寫 Likelihood 方法

In [20]:
from thinkbayes import Suite


class Dice(Suite):
    
    def Likelihood(self, data, hypo):
        if data > hypo: # 如果目前擲到的點數大於假設的骰子面數，則機率為 0.0。例如擲到8點，那6面骰子出現8點的機率為0.0
            return 0.0
        
        return 1.0/hypo # 假設是公平的骰子，每個點數出現的機率是相同的

# 因為目前有4面、6面、8面、12面以及20面的骰子
# 這邊直接用數字來表示假設
suite = Dice([4, 6, 8, 12, 20])

print("還沒有投擲骰子之前的每個骰子的機率分佈")
suite.Print()
suite.Update(6) # 投擲出6點
print()
print("投擲骰子看到6點後的每個骰子的機率分佈")
suite.Print()

print("出現6點以後，投擲的骰子為4面摋機率為0.0")
print("最有可能是來自6面骰子，但幾乎還有12%的機會來自20面骰子")


# 如果我們將抽出來的骰子繼續投擲多次呢？

rolls = [6, 8, 7, 7, 5, 4] # 投擲6次，並且其對應的點數

for data in rolls:
    print()
    suite.Update(data)
    print("投擲骰子看到", data,"點後的每個骰子的機率分佈")
    suite.Print()

還沒有投擲骰子之前的每個骰子的機率分佈
4 0.2
6 0.2
8 0.2
12 0.2
20 0.2

投擲骰子看到6點後的每個骰子的機率分佈
4 0.0
6 0.3921568627450981
8 0.29411764705882354
12 0.19607843137254904
20 0.11764705882352944
出現6點以後，投擲的骰子為4面摋機率為0.0
最有可能是來自6面骰子，但幾乎還有12%的機會來自20面骰子

投擲骰子看到 6 點後的每個骰子的機率分佈
4 0.0
6 0.5256241787122208
8 0.2956636005256242
12 0.1314060446780552
20 0.047306176084099885

投擲骰子看到 8 點後的每個骰子的機率分佈
4 0.0
6 0.0
8 0.7351339577434111
12 0.2178174689610107
20 0.04704857329557832

投擲骰子看到 7 點後的每個骰子的機率分佈
4 0.0
6 0.0
8 0.817574005587765
12 0.16149609986918817
20 0.020929894543046793

投擲骰子看到 7 點後的每個骰子的機率分佈
4 0.0
6 0.0
8 0.8757125344951456
12 0.11532016915162413
20 0.008967296353230295

投擲骰子看到 5 點後的每個骰子的機率分佈
4 0.0
6 0.0
8 0.9158452719690099
12 0.08040342579700499
20 0.0037513022339850668

投擲骰子看到 4 點後的每個骰子的機率分佈
4 0.0
6 0.0
8 0.9432484536722124
12 0.0552061280612909
20 0.001545418266496554


最後有約94%的機會我們是用8面骰子擲出、少於1%是20面骰子。

骰子問題是作者在 Sanjoy Mahajan 的貝葉斯推論的課程上看到的。

你可以在這下載作者的程式碼：http://www.greenteapress.com/thinkbayes/dice.py

#### 請注意程式碼的例子跟書中的例子有不同，所以執行結果不會是上方看到的結果。